### ISSUE TITLE
Frozen dataclass raise error with Protocol with read property

### ISSUE THREAD

**Issue by KerberosMorphy** [microsoft/pylance-release#1813]
## Environment data

-   Language Server version: Pylance v2021.9.2-pre.1
-   VSCode version: 1.60.0
-   OS and version: Windows_NT x64 10.22000.184.0
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.9.7

## Expected behaviour

With a frozen dataclas as:

```python
@dataclass(frozen=True)
class MyDataclass:
    name: str
```

If I design a Protocol using a property decorator (so read only):

```python
class Named(Protocol):
    @property
    def name(self) -> str:
        ...
```
PyLance should act as mypy and **not** raise and error.

And if I use a Protocol with an attribut (so read-write):

```python
class Named(Protocol):
    name: str
```

PyLance should act as mypy **and raise** an error.

## Actual behaviour

PyLance raise an error on a frozen dataclass with a Protocol using property decorator.
PyLance do not raise an error on a frozen dataclass with a Protocol using attribut.

## Code Snippet / Additional information

```python
from __future__ import annotations

from collections.abc import Iterable
from typing import Protocol, TypeVar
from dataclasses import dataclass

class Named(Protocol):
    @property
    def name(self) -> str:
        ...

@dataclass(frozen=True)
class MyDataclass:
    name: str

T = TypeVar("T", bound=Named)


def get_config_from_name(name: str, configs: Iterable[T]) -> T:
    filtered_configs = filter(lambda config: config.name == name, configs)
    config = next(filtered_configs, None)

    if config is None:
        raise LookupError(f"Config doesn't exist for {name}")

    return config

configs = [
    MyDataclass(name="potato")
]
configs = get_config_from_name("potato", configs)  # PyLance Raise Error
```

Error From PyLance:
```
Argument of type "list[MyDataclass]" cannot be assigned to parameter "configs" of type "Iterable[T@get_config_from_name]" in function "get_config_from_name"
  TypeVar "_T_co@Iterable" is covariant
    Type "Named" cannot be assigned to type "MyDataclass"
```

**Comment by erictraut** [comment]
The current behavior is intended.

A property object has semantics that are similar to an attribute, but they are not the same. For this reason, pyright (the type checker upon which pylance is built) does not consider an attribute to be interchangeable with a property. If your protocol specifies a property, a class that conforms to that protocol must also specify a property. If your protocol specifies an attribute, a class that conforms to that protocol must also specify an attribute.

Mypy is "looser" in this respect, but this can lead to type problems. We've chosen to be stricter in this regard.

**Comment by alex007sirois** [comment]
> Mypy is "looser" in this respect, but this can lead to type problems. We've chosen to be stricter in this regard.
Mypy is not exactly just looser.

It's just that a property in a protocol means a read-only attribute for Mypy, and a simple attribute means a read-writable attribute.
The advantage is to be able to specify whether an attribute is writable or not, regardless on whether or not it's a property or attribute.
This means that a frozen dataclass cannot be used with a writable attribute as it would raise an error and also that a function cannot modify a read-only attribute, which is an interesting concept.
To be fair, having a property in a Protocol as well as it's setter is equivalent to a simple attribute, so it is looser in this regards.

Using VSCode for coding and Mypy in the CI pipeline means that code should be compatible with both Pylance/Pyright and Mypy, but read-only protocol attributes with frozen dataclass is one thing that is impossible to reconcile between the two.

I understand the reasoning between keeping simple attributes and property separate, although I disagree. I personally think that it should simply be an implementation detail and using this notation to distinguish between read-only and read-writable attributes is a useful feature.

**Comment by erictraut** [comment]
There are many differences between mypy and pyright. We try to maintain some level of compatibility where it makes sense, but 100% compatibility has never been a goal. I don't think it's feasible to expect that you could see the same type checking results between mypy and pyright. 

In this particular case, pyright more thoroughly tracks whether class-scoped symbols are class or instance variables; mypy largely treats them the same. This allows pyright to avoid certain false positives and catch a number of runtime errors that mypy does not detect. But it also means that pyright needs to treat properties differently from attributes. When you access a property from a class, you get back a property object, whereas when you access an attribute from a class, you either get a runtime error or the value of the attribute depending on whether it's an instance or class variable.

I'll give you a sense for the types of errors that mypy does not catch because of this particular design choice. The following piece of code type checks without errors in mypy.

```python
from typing import ClassVar, Protocol, Type

class FooProtocol(Protocol):
    bar1: ClassVar[int]
    bar2: int
    bar3: int

class Foo:
    bar1: ClassVar[int] = 0

    def __init__(self):
        self.bar2: int = 0

    @property
    def bar3(self) -> int:
        return 0

    @bar3.setter
    def bar3(self, val: int) -> None:
        pass

Foo.bar1
Foo.bar2  # pyright reports error, mypy does not, runtime error
Foo.bar3

f1: Type[FooProtocol] = Foo  # pyright reports error, mypy does not

f1.bar2 = 1  # shouldn't be allowed because bar2 is instance variable
f1.bar3 = 1  # shouldn't be allowed because bar3 is not a class property

f2: FooProtocol = Foo()  # pyright reports error, mypy does not

del f2.bar1  # shouldn't be allowed because bar1 is class variable, runtime error
del f2.bar2
del f2.bar3  # shouldn't be allowed because bar3 has no deleter, runtime error
```

I don't know that it's worth debating this particular design choice. My point is that if your expectation is that pyright (or pylance) will always produce the same type checking results as mypy, I think you're going to be disappointed. We are open to making changes to increase compatibility where it makes sense, but in cases where there are good reasons for the difference in behavior, we're going to be reluctant to make such a change.

If you want to get the same type checking results between VS Code and your CI pipeline, I suggest sticking with the same type checker in both cases. That means one of the following:
1. Use pylance only for language server features. Set typeCheckingMode to "off" so it emits no type errors. Then enable mypy within the Python extension so it is invoked on file save and its errors are reported in the editor.
2. Use pylance for type checking and use the [CLI version of pyright in your CI pipeline](https://github.com/microsoft/pyright/blob/main/docs/ci-integration.md). This is what my team does for our python code base, it it has worked great for us.

If you try either of these approaches, I'd be interested in hearing whether that meets your needs.

**Comment by alex007sirois** [comment]
I understand more clearly why you say Mypy is very loose regarding Protocols. I would consider those bugs, almost. I do agree that Pyright being more severe in the case of ClassVar is a better approach.

I understand that a property will act different than a normal instance variable when targeting the class. However, it is very frequent to have a class switch from normal attribute to a property when more complex behaviors are added. This was the use-case for property in my mind: changing from attribute to functions without altering Interface. However, with the current Pyright implementation, this is not the case since property and instance variables do not share the same interface. This also makes protocol a lot less useful, since different implementation of the Protocol frequently implement the attribute differently: some with normal attribute, some with property when more complex behavior is needed. An interface is there to prevent breaking change in the future, but different behavior of property prevents that.

One solution that makes sense for me would be to make a normal attribute act the same as a property with setter and deletter on an instance, but be absent on a class. Now, the only difference between them would be with `Type[Named]` for example. I'm not sure how that could be perfectly solved, sadly. Honestly, I'm not sure you should be touching the property object of a `Type[Named]`. It seems to me that accessing/modifying a property of a class that is a Type of Protocol is a smell and should not be done, where as being able to specify whether an attribute should be allowed to be written or deleted seems a way more useful feature (to me). An other bonus feature of property are the pydoc that can be added to them and used by IDEs.

To recap, I think normal attributes should act exactly like a read-write-delete property and that property should not be used from a `Type[NamedProtocol]`. This adds useful features, without removing any real use-case (IMO).

I also don't expect Pyright to stay 100% compatible with Mypy. Then, it would have no reason to exist. Switching fully to Mypy or Pyright/Pylance would be the most logical step. Pylance is very well interated into VSCode and it would be a shame to remove it's type-checking abilities. And mypy for CI is a standard. I'll need to look into this to find the best solution.

**Comment by MC-DeltaT** [comment]
I'll add to what @alex007sirois said in that this property vs. attribute problem keeps coming up for me and I'm not sure Pylance's behaviour is ideal. Pyright may be stricter than mypy - which I agree is a good thing, hence why I don't use mypy - but the reality is that the effect of the strictness doesn't match what I, as a human, expect.

I want to be able to express a protocol or abstract base class that allows implementations to provide either an attribute or a property. Maybe I'm just crazy, but this issue comes up all the time when I write Python. It's exacerbated when you take into account dataclasses, which I think are extremely useful, but give no choice of attribute vs. property. It seems like it is impossible to have a nice, safe, type checked solution with Pylance.

What's annoying is that not only does Pylance not recognise that an attribute implements a property interface, but it will give errors if you try to work around the issue. Whichever way you go about it, you get at least 1 error somewhere:

```python
class B1(Protocol):
    foo: int


class I1(B1):
    # Error!
    # "foo" overrides symbol of same name in class "B1"
    #   "property" is incompatible with "int"
    @property
    def foo(self) -> int:
        return 42


class B2(Protocol):
    @property
    def foo(self) -> int:
        raise NotImplementedError()


class I2A(B2):
    def __init__(self) -> None:
        # Error!
        # Cannot assign member "foo" for type "I2A"
        #   Property "foo" has no defined setter
        self.foo = 42


@dataclass(frozen=True)
class I2B(B2):
    # Error!
    # "foo" incorrectly overrides property of same name in class "B2"
    foo: int


class B3(ABC):
    foo: int


class I3(B3):
    # Error!
    # "foo" overrides symbol of same name in class "B3"
    #   "property" is incompatible with "int"
    @property
    def foo(self) -> int:
        return 42
```

In my opinion, it would be nice if the above code with `I2A` and `I2B` were allowed. Maybe there is some technical difference between read-only properties and read-only attributes, but I'd argue in almost all use cases the code is OK.

**Comment by erictraut** [comment]
It looks like you have enabled the strict-mode checks `reportIncompatibleMethodOverride` and `reportIncompatibleVariableOverride`. These are not enabled by default in basic type checking mode. They're intended for developers who want stricter checks for conditions that could lead to potential bugs and crashes. If you don't enable these strict-mode checks, you will see only one error in your code sample — specifically in the I2A class. That's clearly a type violation because `foo` is declared in the base class `B2` as being a read-only property.

Mypy produces the same errors as pyright in your code sample.

[PEP 544](https://peps.python.org/pep-0544/) allows a protocol class to specify a property (with or without a setter), and a class that implements an attribute of the same name should be considered compatible with that protocol. So the following type checks without error in both pyright and mypy.

```python
class Proto1(Protocol):
    @property
    def writable(self) -> int:
        ...

    @writable.setter
    def writable(self, val: int) -> None:
        ...

    @property
    def read_only(self) -> int:
        ...

class Class1:
    writable: int
    read_only: int

c1: Proto1 = Class1()
```

Note that this doesn't work if `Class1` derives from `Proto1`. If `Proto1` is included in the base class list for `Class1`, then the latter is now inheriting the implementation of the former and is subject to any "incompatible override" strict-mode rules that you have enabled.

In your example above, if you change I2B so it doesn't derive from the protocol class, you will not see any incompatible override error.

```python
@dataclass(frozen=True)
class I2B:
    foo: int

i2b: B2 = I2B(2)
```
